### DISCUSSION TITLE
How to implement a nested sort?

### DISCUSSION THREAD

**Discussion by IT-Cru** [spatie/elasticsearch-query-builder#40] [Q&A]
Has anyone managed to implement a nested sort via spatie/elasticsearch-query-builder?

https://www.elastic.co/guide/en/elasticsearch/reference/8.13/sort-search-results.html#nested-sorting

We previously used ONGR\ElasticsearchDSL (https://github.com/ongr-io/ElasticsearchDSL) where it was possible to add nested filtering to a sort, which seems to be not provided by spatie/elasticsearch-query-builder yet.

**Comment by IT-Cru** [comment]
We have now created a NestedSort class in our own code base to get our previous implemented nested sort working again, but I think it would be also nice for others to implement Elastic nested sort and also other possible sort stuff like geo location into spatie/elasticsearch-query-builder itself.

```
<?php

namespace App\Service\Elastic\Sort;

use Spatie\ElasticsearchQueryBuilder\Queries\Query;
use Spatie\ElasticsearchQueryBuilder\Sorts\Sort;

class NestedSort extends Sort
{
    protected string $field;
    protected string $order;

    protected string $path;

    protected ?Query $filter = null;
    protected ?string $missing = null;

    protected ?string $unmappedType = null;

    public function __construct(string $field, string $order)
    {
        parent::__construct($field, $order);
    }

    public static function create(string $field, string $order = Sort::DESC): static
    {
        return new self($field, $order, '');
    }

    public function filter(Query $filter): self
    {
        $this->filter = $filter;

        return $this;
    }

    public function path(string $path): self
    {
        $this->path = $path;

        return $this;
    }

    public function toArray(): array
    {
        $payload = [
            'order' => $this->order,
            'nested' => [
                'path' => $this->path,
            ],
        ];

        if ($this->filter) {
            $payload['nested']['filter'] = $this->filter->toArray();
        }

        return [
            $this->field => $payload,
        ];
    }

}
```

With this NestedSort class we could do now something like this to build our sorting, which is not an ideal solution, because it would be better to have a Sort interface which is used by all possible Elastic sort implementations.

```
$nestedSorting = new NestedSort('prices.value', Sort::DESC);
$nestedSorting->path('prices');
$nestedSorting->filter(new TermQuery('prices.type', 'listing'));

return $nestedSorting;
```

**Comment by dam-bal** [comment]
This discussion can now be closed as `NestedSort` is now part of [v3](https://github.com/spatie/elasticsearch-query-builder/releases/tag/3.0.0) release

**Comment by IT-Cru** [comment]
Works like a charm. After upgrading from 1.x to 3.x it was possible to remove our custom NestedSort code in our project.
